oauth_samples/go/oauth.go

```
@@ -0,0 +1,155 @@
+	// user:read,workspaces:read
+	scopes = "<<enter the scopes your application requires>>"
+	state  = uuid.New().String()
+
+	// client (application) credentials - located at https://twist.com/integrations/build, then go to your integration
+	clientID     = os.Getenv("twist_clientID")
+	clientSecret = os.Getenv("twist_clientSecret")
+)
+
+func main() {
+	if err := requestAuthorization(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func requestAuthorization() error {
+	reader := bufio.NewReader(os.Stdin)
+
+	// Step 1 - simulate a request from a browser on the authorize_url -
+	// will return an authorization code after the user is prompted for credentials.
+	vals := make(url.Values)
+	vals.Set("response_type", "code")
+	vals.Set("client_id", clientID)
+	vals.Set("redirect_url", callbackURL)
+	vals.Set("scope", scopes)
+	vals.Set("state", state)
+
+	fmt.Println("Go to the following url on a browser and enter the code from the returned url")
+	fmt.Println(authorizeURL + "?" + vals.Encode())
+	fmt.Printf("-> ")
+
+	// Read the authorization code
+	text, readerError := reader.ReadString('\n')
+
+	if readerError != nil {
+		fmt.Println("Erorr reading the code you entered")
+		return readerError
+	}
+
+	requestToken(strings.TrimSuffix(text, "\n"))
+
+	return nil
+}
+
+func requestToken(code string) error {
+	fmt.Println("Requesting access token")
+
+	// Step 2, 3 - turn the authorization code into a access token, etc
+	token, err := exchangeForToken(code)
+
+	if err != nil {
+		fmt.Println("Invalid token response")
+		return err
+	}
+
+	// We can now use the access_token as much as we want to access protected resources.
+	return makeAPICall(token)
+}
+
+func makeAPICall(token string) error {
+	var hc = http.DefaultClient
+	var bearer = "Bearer " + token
+
+	request, requestError := http.NewRequest("GET", apiURL, nil)
+
+	if requestError != nil {
+		fmt.Println("Error creating request")
+		return requestError
```

Better to return an annotated error here instead of print + return.